### ISSUE TITLE
"Envoy may have been out of memory killed. Check memory usage and limits."

### ISSUE THREAD

**Issue by Stono** [istio/istio#19925]
**Bug description**
I'm seeing a sidecar for one of our applications continually have its memory grow and eventually OOM:

```
2020-01-02T14:44:33.668963Z     info    Envoy proxy is ready
2020-01-04T15:26:26.987721Z     warn    Envoy may have been out of memory killed. Check memory usage and limits.
2020-01-04T15:26:26.987766Z     error   Epoch 0 exited with error: signal: killed
2020-01-04T15:26:26.987770Z     info    No more active epochs, terminating
```

You can see the memory growth in this graph:
<img width="795" alt="Screen Shot 2020-01-05 at 15 28 24" src="https://user-images.githubusercontent.com/1486729/71782357-f77ba500-2fd0-11ea-846f-291b6d7b50c4.png">

Typical sidecar usage is about 50mb.

It's consistent, so i'm confident it's something unique about this particular application that's causing it - but can't figure out what, I'm just not sure where to begin with debugging it - so looking for some advice please.

**Expected behavior**
Envoy to not OOM

**Steps to reproduce the bug**
Can't give them yet

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.4.2

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE 1.14

**Comment by Stono** [comment]
Hmm actually now that i'm digging into this further, i'm seeing it in other places:

<img width="800" alt="Screen Shot 2020-01-05 at 15 59 41" src="https://user-images.githubusercontent.com/1486729/71782645-732b2100-2fd4-11ea-8fad-c6d72d6dd301.png">

<img width="797" alt="Screen Shot 2020-01-05 at 16 00 35" src="https://user-images.githubusercontent.com/1486729/71782650-86d68780-2fd4-11ea-9af3-f4b01cf74678.png">

**Comment by howardjohn** [comment]
Interesting how consistent it is. Seems like the times it spikes up are somewhat correlated with time of day, maybe due to traffic patterns or cronjob, etc.

https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#generate-profile maybe able to isolate what code path is using the memory

**Comment by howardjohn** [comment]
have you been on 1.4.2 for a while without this issue, or is it potentially a 1.4 vs 1.3 issue

**Comment by Stono** [comment]
Consistent on that one app yeah.  

Honestly I have no idea when it started as unfortunately until recently our alert was for “FrequentlyRestartingContainers” and this didn’t qualify as frequently.  I added an alert for any restart recently which has highlighted it to me on the first app I mentioned.  I then just went digging for proxied using notably more memory than their request.

**Comment by Stono** [comment]
I could run a 1.3.5 sidecar on that app and see what happens?  What do you think?

**Comment by howardjohn** [comment]
Seems like a good idea if it's feasible. I personally always try to narrow down variables when trying to figure out issues like this since there are a lot of possible root causes and that would narrow it down quite a bit

**Comment by Stono** [comment]
It is.  Will report back

**Comment by Stono** [comment]
Right the app in my opening post which is pretty regular over a 24h period is on `1.3.5`, will let you know what i see tomorrow evening.

Cheers

**Comment by Stono** [comment]
<img width="797" alt="Screenshot 2020-01-06 at 07 38 51" src="https://user-images.githubusercontent.com/1486729/71803050-9f3bb600-3057-11ea-9842-aae8787d6bbe.png">

Looking suspiciously like a `1.4` issue @howardjohn - the above graph is since switching back to a `1.3.5` proxy

**Comment by howardjohn** [comment]
@lambdai did we change anything in the proxy that could trigger this?

**Comment by Stono** [comment]
Spoke to soon.  Jumped up not longer after after I posted that comment.

<img width="802" alt="Screenshot 2020-01-06 at 15 11 00" src="https://user-images.githubusercontent.com/1486729/71826817-c82f6b80-3096-11ea-832f-79f3bdadf380.png">

So doesn't look like a proxy `1.4` thing

**Comment by howardjohn** [comment]
The fact it goes so long without issues makes me think it may be from the config/xds code path. Perhaps something changes in the cluster, triggers an update, and then Envoy holds onto some old data structures or something? Getting the output of /memory on the envoy may give a little more info about if its allocated or just heap memory usage

**Comment by Stono** [comment]
Yeah I agree but I couldn’t correlate it to a specific event, albeit I’ve been dipping in and out of this so will have a deeper look tomorrow.

One of the guys on my team is making some changes to our injector config now so that we are able to remove the read only root, to enable us to heap dump.

It’s definitely triggered by something, just no idea what!  I’ll turn on more verbose envoy logs for this app too and see if I see anything obvious

**Comment by Stono** [comment]
OK @howardjohn so I've managed to correlate the spikes to some proxy logs that are interesting:

```
[Envoy (Epoch 0)] [2020-01-07 08:03:51.210][21][debug][config] [external/envoy/source/common/config/grpc_mux_subscription_impl.cc:70] gRPC update for type.googleapis.com/envoy.api.v2.RouteConfiguration failed
[Envoy (Epoch 0)] [2020-01-07 08:03:51.210][21][debug][config] [external/envoy/source/common/config/grpc_mux_subscription_impl.cc:70] gRPC update for type.googleapis.com/envoy.api.v2.ClusterLoadAssignment failed
```

<img width="764" alt="Screenshot 2020-01-07 at 08 04 52" src="https://user-images.githubusercontent.com/1486729/71878788-6cabbf00-3124-11ea-81f2-0d95def23b02.png">

Full log attached.

[proxy-logs.txt](https://github.com/istio/istio/files/4029470/proxy-logs.txt)

**Comment by Stono** [comment]
Also I just stumbled across this, if it helps you narrow things down.  Mixer (not the proxy, mixer itself) did the same jump a few days ago:

<img width="917" alt="Screenshot 2020-01-07 at 10 34 04" src="https://user-images.githubusercontent.com/1486729/71889093-53613d80-3139-11ea-8728-e6b4a9904a02.png">

**Comment by howardjohn** [comment]
Interesting its mixer itself. I can't think of any scenario that would cause both Envoy and Mixer to have increased memory usage, so that is pretty strange.. the only thing I can think of for that would be https://github.com/istio/istio/issues/19481 which is fixed in 1.4.3

So based on the logs, we close the XDS connection exactly 5 minutes after the last update (so likely from keepalive), then reconnect, and immediately the memory spikes. So looks like its definitely on the config path.

Is it correct that its consistent in this application but rarely happening in others? Is there anything different about this one? Are others in the same namespace/similar `Sidecar` scoping experiencing the same? Just wondering if its a specific type of config that may be triggering this.

Running a diff on the config_dump before and after may be useful to see if its getting the same config and behaving bad, or getting different config and behaving badly. Ill ask @lambdai what other debugging we can do, we are pushing the limits of my envoy expertise

**Comment by Stono** [comment]
Hey, 
So it's happening on some (about 10 that I've spotted, from 320).  They're all deployed identically as we have standardised application deployments.  Their Sidecars are only configured with their dependent services which vary obviously but other than that, they're identical.  I wasn't able to draw any correlation between egress routes.

I have someone on my team looking at using the memory profiling however it's proving difficult as you don't seem to be able to just capture a heap dump at a point in time (someone with more experience might be able to tell us other wise).

**Comment by lambdai** [comment]
Based on what you described so far, the memory usage is highly like originate from update procedure.
Some ideas:
1. proto messages: Pilot will push the latest XDS config to envoy. No matter whether envoy would digest them. Those message could be left in grpc buffer which is 256 MB by default. The stats of xds cluster bytes transmitted and number LDS/CDS/EDS update could be the estimation. If so , try apply a small number and leave the back pressure in pilot :)

**Comment by lambdai** [comment]
2. Too many update: There are multiple version tracked in envoy runtime. Including draining listener, Snapped RDS until the upstream connection is gone, similar to EDS which I don't have a clear sense.
I don't have a solution for this though

**Comment by cinder-fish** [comment]
Hi, I work in @Stono's team, we've observed this happening again for the same application and this time have captured some debug output from `/memory` `/stats` and `/config_dump` at points before and after the memory increase (attached)

<img width="706" alt="Screen Shot 2020-01-08 at 09 59 15" src="https://user-images.githubusercontent.com/2643547/71969084-e663ab80-31fd-11ea-9ef5-e702fb8d5aa2.png">

[envoy-debugging-samples.tar.gz](https://github.com/istio/istio/files/4034680/envoy-debugging-samples.tar.gz)

**Comment by cinder-fish** [comment]
There's another pod from the same replicaset whose envoy was OOM killed at the same time (2020-01-07T22:10:58Z)

**Comment by howardjohn** [comment]
Ok so based on the above, the config dump is the same. The heap size jumped up but allocated remains the same. @lambdai

**Comment by Stono** [comment]
Any ideas folks?  This is one of the main causes of memory contention on our nodes right now.  I could set the limit for the containers lower and they'll just OOM whenever this happens, but that causes service impact momentarily.

**Comment by Stono** [comment]
Here's my latest favourite, 3 different pods (nginx) sidecar, on 3 different nodes, at exactly the same time.

<img width="801" alt="Screenshot 2020-01-10 at 11 16 47" src="https://user-images.githubusercontent.com/1486729/72149322-c7e6e700-339a-11ea-862b-5cef92d9c3e9.png">
<img width="798" alt="Screenshot 2020-01-10 at 11 16 57" src="https://user-images.githubusercontent.com/1486729/72149323-c7e6e700-339a-11ea-959d-7d226dc9df9e.png">
<img width="799" alt="Screenshot 2020-01-10 at 11 17 05" src="https://user-images.githubusercontent.com/1486729/72149324-c7e6e700-339a-11ea-8db5-606d73748272.png">

**Comment by Stono** [comment]
They all seem to jump approximately 3.5x

**Comment by Stono** [comment]
Also as you can see 650mb is far more than the 256mb buffer @lambdai mentioned, so it kind of discounts that

**Comment by lambdai** [comment]
I compared the two stats in the compress file
There are very few signals: the major diffs originates from the life time of server
```
  server.uptime: 28703                                                                                                 
  server.uptime: 89491                                                                                                 
```
```
  cluster_manager.cds.update_attempt: 191                                                                              |  cluster_manager.cds.update_attempt: 586                                                                              
  listener_manager.lds.update_attempt: 191                                                                             |  listener_manager.lds.update_attempt: 586                                                                             
```
Is there any other worth noticing stuff during the mem usage rocketing period?
What does the nginx say at that time? Is there a req spike?

**Comment by howardjohn** [comment]
@lambdai I think from the logs its correlated with a new connection to pilot

**Comment by lambdai** [comment]
@howardjohn  Good point! The reset every 30 min.
So the concerning xds update is hidden while we hunt for the abnormal update counts.

**Comment by Stono** [comment]
Is there anything else we can do to help here guys?  This is causing quite a bit of node contention for us.

This prom query shows the % memory (vs request) of sidecars > 100%. 

<img width="1670" alt="Screenshot 2020-01-13 at 11 50 58" src="https://user-images.githubusercontent.com/1486729/72254110-04188280-35fb-11ea-9d8d-69087b2c7e05.png">

**Comment by Stono** [comment]
Hi @howardjohn @lambdai - I've found something else which might help?

This is the flow:

`[sauron-web]` -> `[lead-generation-service]` -> `[search-one]`

Take this big spike:
<img width="801" alt="Screenshot 2020-02-02 at 12 23 49" src="https://user-images.githubusercontent.com/1486729/73608160-35abba80-45b7-11ea-8aee-94142ecc6098.png">

It correlates in a jump in latency on that service (actually because `search-one` was running slow)
<img width="796" alt="Screenshot 2020-02-02 at 12 24 01" src="https://user-images.githubusercontent.com/1486729/73608162-3e9c8c00-45b7-11ea-8540-b42aabaa61fb.png">

`sauron-web` in this example has a timeout which it kills the connection to `lead-generation-service` after 1 second.  This manifests in istio as a 500 error with `response_flags` `DC`:

<img width="1667" alt="Screenshot 2020-02-02 at 12 25 38" src="https://user-images.githubusercontent.com/1486729/73608181-83282780-45b7-11ea-844d-96b503e05592.png">

Notice how the spikes in proxy memory correlate exactly with the spikes in DCs?

**Comment by Stono** [comment]
this is still really hurting us :-(

the only correlation i can draw so far is that it coincides with latency in the downstream service (causing a latency spike in this service, which results in some `500 DC` errors due to the upstream killing the connection due to timeouts being hit.

<img width="801" alt="Screenshot 2020-02-02 at 12 23 49" src="https://user-images.githubusercontent.com/1486729/73609884-c2f80a80-45c9-11ea-8d35-ddf5b29c4310.png">

If i had to shoot in the dark, something about a `500, DC` scenario is leaving memory allocated in envoy

**Comment by yxue** [comment]
> I have someone on my team looking at using the memory profiling however it's proving difficult as you don't seem to be able to just capture a dump at a point in time, rather you have to record change over time - and with this happening in a non-deterministic fashion we're unlikely to be able to capture it (someone with more experience might be able to tell us other wise).

I agree that capturing the moment of the memory change would be helpful for diagnostics but it's hard to get that. I am wondering if we can get the snapshot of the memory when its usage is high, it would help us to figure out which objects are kept in the memory. That would be helpful if the issue is raised by the memory leak of the Envoy.

**Comment by yxue** [comment]
`update_failure` from the stats in the zip file

```
cluster_manager.cds.update_attempt: 191
cluster_manager.cds.update_failure: 90
cluster_manager.cds.update_rejected: 0
cluster_manager.cds.update_success: 100
listener_manager.lds.update_attempt: 191
listener_manager.lds.update_failure: 90
listener_manager.lds.update_rejected: 0
listener_manager.lds.update_success: 100
```

LDS/CDS update failed due to the connection failure and the failure ratio is ~50%. I am not sure if this is expected? @howardjohn

**Comment by duderino** [comment]
> `update_failure` from the stats in the zip file
> 
> ```
> cluster_manager.cds.update_attempt: 191
> cluster_manager.cds.update_failure: 90
> cluster_manager.cds.update_rejected: 0
> cluster_manager.cds.update_success: 100
> listener_manager.lds.update_attempt: 191
> listener_manager.lds.update_failure: 90
> listener_manager.lds.update_rejected: 0
> listener_manager.lds.update_success: 100
> ```
> 
> LDS/CDS update failed due to the connection failure and the failure ratio is ~50%. I am not sure if this is expected? @howardjohn

Can we compare those metrics to a good node that has been running for a comparably long time?

**Comment by Stono** [comment]
Sure, this one has been running for 8 days and is very normal memory:

```
envoy_cluster_manager_cds_update_attempt{} 4735
envoy_cluster_manager_cds_update_failure{} 2271
envoy_cluster_manager_cds_update_rejected{} 0
envoy_cluster_manager_cds_update_success{} 2463
envoy_listener_manager_lds_update_attempt{} 4735
envoy_listener_manager_lds_update_failure{} 2271
envoy_listener_manager_lds_update_rejected{} 0
envoy_listener_manager_lds_update_success{} 2463
```

<img width="1652" alt="Screenshot 2020-02-05 at 20 30 15" src="https://user-images.githubusercontent.com/1486729/73880442-59406080-4856-11ea-9b57-a6fba725c730.png">

**Comment by yxue** [comment]
Looks like the 50% `update_failure` is expected due to the gRPC reconnection between pilot and envoy. 

One hypothesize is the xDS protobuf mem leak. If we can get the memory dump and figure out objects in the memory. We might be able to narrow the investigate scope. 

cc @jplevyak who is investigating another high memory issue.

**Comment by Stono** [comment]
If it was a generic xDS leak, we'd surely see it across a broader range of services?

**Comment by Stono** [comment]
<img width="1618" alt="Screenshot 2020-02-11 at 08 16 09" src="https://user-images.githubusercontent.com/1486729/74220196-d29ce080-4ca6-11ea-86f4-bf3d2763d432.png">

:(

**Comment by yxue** [comment]
From the stats and config dump, I didn't find any evidence pointing to the the place of memory leak. Repeat what was mentioned in the thread, heap memory increased from 

```
server.memory_heap_size: 15728640
```
to 
```
server.memory_heap_size: 200278016
```

I think figuring out what is remaining in the memory (a memory dump when memory usage is high) would help us to diagnostics the memory leak.

**Comment by Stono** [comment]
Is there any way to get a heap dump at a point in time?
It's really not realistic we can attach a memory profile to an application for days (as sometimes it can take a few days for the memory to do its jump)

**Comment by Stono** [comment]
It's happened in pilot too - so not just proxies.

<img width="1676" alt="Screenshot 2020-02-16 at 21 12 58" src="https://user-images.githubusercontent.com/1486729/74612840-283b1800-5101-11ea-9b10-f506ff9540ef.png">

**Comment by howardjohn** [comment]
Pilot has a lot of different memory stats exposed, may give some more info there. You can see them in the default pilot dashboard.

**Comment by Stono** [comment]
wow these are some worrying graphs, pilot over the last 14 days (resident and virtual memory are jumping consistently):

<img width="1655" alt="Screenshot 2020-02-16 at 21 26 13" src="https://user-images.githubusercontent.com/1486729/74613087-312ce900-5103-11ea-8bf7-251fe268a9da.png">

**Comment by duderino** [comment]
> From the stats and config dump, I didn't find any evidence pointing to the the place of memory leak. Repeat what was mentioned in the thread, heap memory increased from
> 
> ```
> server.memory_heap_size: 15728640
> ```
> 
> to
> 
> ```
> server.memory_heap_size: 200278016
> ```
> 
> I think figuring out what is remaining in the memory (a memory dump when memory usage is high) would help us to diagnostics the memory leak.

I just want to call attention to this one ^  I asked @yxue to look into this but his investigation is blocked on the heap dump.  If you can give him a few (e.g., a known good one + a few bad ones) that should allow further investigation to continue.

https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#generate-profile

**Comment by Stono** [comment]
It may be too soon to tell, but we've turned of `controlPlaneSecurity` around 48 hours ago due to https://github.com/istio/istio/issues/18090#issuecomment-586863034, and as of yet haven't had a single sidecar breach it's memory.

**Comment by lambdai** [comment]
Do we all agree there is no solid leak? There was a peak heap usage but eventually envoy return the memory to malloc.
I can see 2 problems here:
1. tcmalloc doesn't return to OS. It's a known issue at low priority. It hurts us many times so I will take it.
2. Envoy doesn't protect itself when pilot floods new configs. What envoy is doing is execute the update one by one, as fast as envoy can. In the mean time, temporary memory usage is high.

It's not that hard to explain that adding another envoy between pilot and sidecar surfaces the problem: It's introducing another buffer which could flood the sidecar. The memory usage at pilot container also reveals the buffer abusing.
The throttling better to be applied at every layer.
1. envoy inbound buffer for xds connection. Low hanging fruit
2. Throttling at envoy config applying: It should work in most of the cases since the config execution is always taking much more memory than the serialized config as a protobuf obj.
3. Configs group commit: During the flood we expect multiple version of config is given and we care about the latest one. 
4. pilot: It will be awesome if pilot could dedup the config. 

Let me draft PR of item 1. It's not ideal but I think it will bring the memory usage back to `controlPlaneSecurity=off`, if not better.

**Comment by Stono** [comment]
Yeah i'm not sure it's a leak per-se, but having an istio-proxy that goes from 60mb to 500mb in a moment (and never goes back) is a huge problem for many operators.  Too many of those on a multi-tenant node and your node goes pop!

**Comment by ramaraochavali** [comment]
@lambdai For #2, you may look at applying ratelimit settings for ads_config source to slow down the config ingestion at Envoy. See https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/config_source.proto#envoy-api-msg-core-apiconfigsource and https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/config_source.proto#core-ratelimitsettings if it helps

**Comment by sachithmuhandiram** [comment]
We faced this OOM in Envoy Side-car

> warn Envoy may have been out of memory killed. Check memory usage and limits.

This happened only for one time. No regular crashes as @Stono has.

We have `Istio-1.7.3` in baremetal Kubernetes cluster (`v1.19.2`)

**Comment by tariq1890** [comment]
> tcmalloc doesn't return to OS. It's a known issue at low priority. It hurts us many times so I will take it.

@lambdai Is this still the case with the more recent versions of Istio?  We are on Istio 1.7.4 and we see issues where the istio-proxy memory consumption is ever-increasing and never dipping. 

Is this the recommended solution setting `TCMALLOC_RELEASE_RATE` envar variable? That's what my understanding is from this Issue : https://github.com/envoyproxy/envoy/issues/7151

**Comment by JonasJasas** [comment]
Having the same issue with the v1.9.3 .The memory leak could be related to the `retries` option. Unable to reproduce.

```
2021-04-23T11:01:48.883711Z    info    ads    ADS: "@" router~172.20.53.93~istio-ingressgateway-6dd6579bd8-q7xmp.istio-system~istio-system.svc.cluster.local-1 terminated rpc error: code = Canceled desc = context canceled                                                                                                                                                                                                           
2021-04-23T11:01:48.883711Z    info    ads    ADS: "@" router~172.20.53.93~istio-ingressgateway-6dd6579bd8-q7xmp.istio-system~istio-system.svc.cluster.local-2 terminated rpc error: code = Canceled desc = context canceled                                                                                                                                                                                                           
2021-04-23T11:01:48.914643Z    warn    Envoy may have been out of memory killed. Check memory usage and limits.                                                                                                     
2021-04-23T11:01:48.914722Z    error    Epoch 0 exited with error: signal: killed                                                                                                                                   
2021-04-23T11:01:48.914730Z    info    No more active epochs, terminating                                                                                                                                           
stream closed
``

**Comment by 15989206052** [comment]
> **Bug description**
> I'm seeing a sidecar for one of our applications continually have its memory grow and eventually OOM:
> 
> ```
> 2020-01-02T14:44:33.668963Z     info    Envoy proxy is ready
> 2020-01-04T15:26:26.987721Z     warn    Envoy may have been out of memory killed. Check memory usage and limits.
> 2020-01-04T15:26:26.987766Z     error   Epoch 0 exited with error: signal: killed
> 2020-01-04T15:26:26.987770Z     info    No more active epochs, terminating
> ```
> 
> You can see the memory growth in this graph:
> <img alt="Screen Shot 2020-01-05 at 15 28 24" width="795" src="https://user-images.githubusercontent.com/1486729/71782357-f77ba500-2fd0-11ea-846f-291b6d7b50c4.png">
> 
> Typical sidecar usage is about 50mb.
> 
> It's consistent, so i'm confident it's something unique about this particular application that's causing it - but can't figure out what, I'm just not sure where to begin with debugging it - so looking for some advice please.
> 
> **Expected behavior**
> Envoy to not OOM
> 
> **Steps to reproduce the bug**
> Can't give them yet
> 
> **Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
> 1.4.2
> 
> **How was Istio installed?**
> Helm
> 
> **Environment where bug was observed (cloud vendor, OS, etc)**
> GKE 1.14
I have encountered the same question. Have you found the answer?

**Comment by howardjohn** [comment]
@15989206052 your version is about 2.5 years old. Please upgrade, no one will help you on this old of a version.

**Comment by dominicqi** [comment]
> Having the same issue with the v1.9.3 .The memory leak could be related to the `retries` option. Unable to reproduce.
> 
> ```
> 2021-04-23T11:01:48.883711Z    info    ads    ADS: "@" router~172.20.53.93~istio-ingressgateway-6dd6579bd8-q7xmp.istio-system~istio-system.svc.cluster.local-1 terminated rpc error: code = Canceled desc = context canceled                                                                                                                                                                                                           
> 2021-04-23T11:01:48.883711Z    info    ads    ADS: "@" router~172.20.53.93~istio-ingressgateway-6dd6579bd8-q7xmp.istio-system~istio-system.svc.cluster.local-2 terminated rpc error: code = Canceled desc = context canceled                                                                                                                                                                                                           
> 2021-04-23T11:01:48.914643Z    warn    Envoy may have been out of memory killed. Check memory usage and limits.                                                                                                     
> 2021-04-23T11:01:48.914722Z    error    Epoch 0 exited with error: signal: killed                                                                                                                                   
> 2021-04-23T11:01:48.914730Z    info    No more active epochs, terminating                                                                                                                                           
> stream closed
> ``
> ```

hi, i have the same problem，does setting TCMALLOC_RELEASE_RATE env works? Thanks.

**Comment by dominicqi** [comment]
> > tcmalloc doesn't return to OS. It's a known issue at low priority. It hurts us many times so I will take it.
> 
> @lambdai Is this still the case with the more recent versions of Istio? We are on Istio 1.7.4 and we see issues where the istio-proxy memory consumption is ever-increasing and never dipping.
> 
> Is this the recommended solution setting `TCMALLOC_RELEASE_RATE` envar variable? That's what my understanding is from this Issue : [envoyproxy/envoy#7151](https://github.com/envoyproxy/envoy/issues/7151)

hi, i have the same problem，does setting TCMALLOC_RELEASE_RATE env works? Thanks.

**Comment by JonasJasas** [comment]
We are using 1.9.5 and constantly having this issue and `TCMALLOC_RELEASE_RATE` makes no difference.

**Comment by 15989206052** [comment]
> @15989206052 your version is about 2.5 years old. Please upgrade, no one will help you on this old of a version.

many thx, I forgot that I described it like this, i always use the almost latest version, by the way, i had resolved it by set MAX_ACTIVES_STREAMS.

**Comment by lmgr10** [comment]
> > @15989206052 your version is about 2.5 years old. Please upgrade, no one will help you on this old of a version.
> 
> many thx, I forgot that I described it like this, i always use the almost latest version, by the way, i has resolved it by limiting MAX_ACTIVES_STREAMS.

Hi! you mean MAX_ACTIVE_STREAMS by max_concurrent_streams in http2_protocol_options?
Which value have you set on that option?

Thx in advance!

**Comment by 15989206052** [comment]
> > > @15989206052 your version is about 2.5 years old. Please upgrade, no one will help you on this old of a version.
> > 
> > 
> > many thx, I forgot that I described it like this, i always use the almost latest version, by the way, i has resolved it by limiting MAX_ACTIVES_STREAMS.
> 
> Hi! you mean MAX_ACTIVE_STREAMS by max_concurrent_streams in http2_protocol_options? Which value have you set on that option?
> 
> Thx in advance!

Sorry for my lateness, it's the peer's MAX_ACTIVES_STREAMS.

**Comment by adviteey1** [comment]
@15989206052 @easverasuthan We are also stuck with the same issue and couldn't find where to change the value suggested?
